### ISSUE TITLE
Map#setLayerSource to allow changing a layer's source

### ISSUE THREAD

**Issue by andrewharvey** [mapbox/mapbox-gl-js#6895]
This would let you change a layer's source. Currently you need to

1. getLayer (due to #3713 you need to getStyle then find the layer)
2. removeLayer
3. addLayer

This would make it easier to to use local GeoJSON sources on the client side together with layers from Mapbox Studio. In Studio you can't create local GeoJSON source layers so you need to use a dummy source, which requires swapping out on the client side.

**Comment by vinayakkulkarni** [comment]
Yep, swapping out geojson can be a pain in the a$$, `setData()` feels kind of a hack.. would really appreciate a way to swap out geojson.. 
maybe something like: `map.setSource('layer-id', <some geosjon>)`
